rustls-cert-gen/src/cert.rs

```diff
@@ -0,0 +1,433 @@
+	/// let cert = CertificateBuilder::new().signature_algorithm("ed25519");
+	/// ```
+	pub fn signature_algorithm(mut self, alg: &str) -> anyhow::Result<Self> {
+		let sig = Signature::new(alg)?;
+		self.params.alg = sig.key_pair.algorithm();
+		self.params.key_pair = Some(sig.key_pair);
+		Ok(self)
+	}
+	/// Set options for Ca Certificates
+	/// # Example
+	/// ```
+	/// # use rustls_cert_gen::CertificateBuilder;
+	/// let cert = CertificateBuilder::new().certificate_authority();
+	/// ```
+	pub fn certificate_authority(self) -> CaParams {
+		CaParams::new(self.params)
+	}
+	/// Set options for `EndEntity` Certificates
+	pub fn end_entity(self) -> EndEntityParams {
+		EndEntityParams::new(self.params)
+	}
+}
+
+pub struct CaParams {
+	params: CertificateParams,
+}
+
+impl CaParams {
+	pub fn new(mut params: CertificateParams) -> Self {
+		params.is_ca = IsCa::Ca(BasicConstraints::Unconstrained);
+		params.key_usages.push(KeyUsagePurpose::DigitalSignature);
+		params.key_usages.push(KeyUsagePurpose::KeyCertSign);
+		params.key_usages.push(KeyUsagePurpose::CrlSign);
+		Self { params }
+	}
+	pub fn country_name(mut self, country: &str) -> Self {
+		self.params
+			.distinguished_name
+			.push(DnType::CountryName, PrintableString(country.into()));
+		self
+	}
+	pub fn organization_name(mut self, name: &str) -> Self {
+		self.params
+			.distinguished_name
+			.push(DnType::OrganizationName, name);
+		self
+	}
+	pub fn build(self) -> Result<Ca, rcgen::Error> {
+		let cert = Certificate::from_params(self.params)?;
+		let cert = Ca { cert };
+		Ok(cert)
+	}
+}
+
+pub struct Ca {
+	cert: Certificate,
+}
+
+impl Ca {
+	/// Self-sign and serialize
+	pub fn serialize_pem(&self) -> Result<PemCertifiedKey, rcgen::Error> {
+		let cert_pem = self.cert.serialize_pem()?;
+		let key_pem = self.cert.serialize_private_key_pem();
+		Ok(PemCertifiedKey {
+			cert_pem,
+			private_key_pem: key_pem,
+		})
+	}
+	pub fn cert(&self) -> &Certificate {
+		&self.cert
+	}
+}
+
+/// End-entity [Certificate]
+pub struct EndEntity {
+	cert: Certificate,
+}
+
+impl EndEntity {
+	/// Sign with `self.signer` and serialize.
+	pub fn serialize_pem(&self, signer: &Certificate) -> Result<PemCertifiedKey, rcgen::Error> {
+		let cert_pem = self.cert.serialize_pem_with_signer(signer)?;
+		let key_pem = self.cert.serialize_private_key_pem();
+		Ok(PemCertifiedKey {
+			cert_pem,
+			private_key_pem: key_pem,
+		})
+	}
+}
+
+/// [CertificateParams] from which an [EndEntity] [Certificate] can be built
+pub struct EndEntityParams {
+	params: CertificateParams,
+}
+
+impl EndEntityParams {
+	/// Initialize `EndEntityParams`
+	pub fn new(mut params: CertificateParams) -> Self {
+		params.is_ca = IsCa::NoCa;
+		params.use_authority_key_identifier_extension = true;
+		params.key_usages.push(KeyUsagePurpose::DigitalSignature);
+		Self { params }
+	}
+	pub fn common_name(mut self, name: &str) -> Self {
+		self.params
+			.distinguished_name
+			.push(DnType::CommonName, name);
+		self
+	}
```

Comment:
It looks like calling this multiple times will append multiple `CommonName` DNs. I think probably this should only set the Common Name DN once? Or it should have a docstring describing the behaviour.

Reply:
I think this is a special case. `push` is a bit misleading since `DistinquishedName` is backed by a `hashmap`. Under the hood, `entries.insert()` is called with the `DnType` as the key. So a second push will replace rather than append. So I think the behavior is ok. We can add a docstring like "Set CommonName. Multiple calls will replace previous value".

Reply:
Ah! Great point. You're right :+1: Sorry about the confusion.